clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java

```diff
@@ -151,236 +104,142 @@ else if (clientAuthConfig.equals("requested"))
+            throw new IllegalStateException("SslFactory has not been configured.");
+        }
+        Map<String, Object> nextConfigs = new HashMap<>(sslEngineBuilder.configs());
+        copyMapEntries(nextConfigs, newConfigs, SslConfigs.RECONFIGURABLE_CONFIGS);
+        if (clientAuthConfigOverride != null) {
+            nextConfigs.put(BrokerSecurityConfigs.SSL_CLIENT_AUTH_CONFIG, clientAuthConfigOverride);
+        }
+        if (!sslEngineBuilder.shouldBeRebuilt(nextConfigs)) {
+            return;
+        }
         try {
-            SecurityStore newKeystore = maybeCreateNewKeystore(configs);
-            SecurityStore newTruststore = maybeCreateNewTruststore(configs);
-            if (newKeystore != null || newTruststore != null) {
-                SecurityStore keystore = newKeystore != null ? newKeystore : this.keystore;
-                SecurityStore truststore = newTruststore != null ? newTruststore : this.truststore;
-                createSSLContext(keystore, truststore);
+            SslEngineBuilder newSslEngineBuilder = new SslEngineBuilder(nextConfigs);
+            if (sslEngineBuilder.keystore() == null) {
+                if (newSslEngineBuilder.keystore() != null) {
+                    throw new ConfigException("Cannot add SSL keystore to an existing listener for " +
+                            "which no keystore was configured.");
+                }
+            } else {
+                if (newSslEngineBuilder.keystore() == null) {
+                    throw new ConfigException("Cannot remove the SSL keystore from an existing listener for " +
+                            "which a keystore was configured.");
+                }
+                if (!CertificateEntries.create(sslEngineBuilder.keystore().load()).equals(
+                        CertificateEntries.create(newSslEngineBuilder.keystore().load()))) {
+                    throw new ConfigException("Keystore DistinguishedName or SubjectAltNames do not match");
+                }
+            }
+            if (sslEngineBuilder.truststore() == null && newSslEngineBuilder.truststore() != null) {
+                throw new ConfigException("Cannot add SSL truststore to an existing listener for which no " +
+                        "truststore was configured.");
+            }
+            if (keystoreVerifiableUsingTruststore) {
+                if (sslEngineBuilder.truststore() != null || sslEngineBuilder.keystore() != null) {
+                    SslEngineValidator.validate(sslEngineBuilder, newSslEngineBuilder);
+                }
             }
         } catch (Exception e) {
-            log.debug("Validation of dynamic config update of SSL keystore/truststore failed", e);
-            throw new ConfigException("Validation of dynamic config update of SSL keystore/truststore failed: " + e);
+            log.info("Unable to validate proposed new SslFactory configuration.", e);
+            throw new ConfigException("Unable to reconfigure SslFactory: " + e);
```

Comment:
This is a validation failure (e.g. admin client validateOnly request), sobetter to say `validation failed`.